votor/src/consensus_pool/vote_pool.rs

```diff
@@ -2,239 +2,375 @@ use {
+                    Entry::Occupied(mut e) => {
+                        let map = e.get_mut();
+                        match map.entry(nf.block_id) {
+                            Entry::Vacant(map_e) => {
+                                map_e.insert(vote);
+                                Ok(())
+                            }
+                            Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                        }
+                    }
+                }
+            }
+            Vote::Skip(_) => {
+                if self.notar.contains_key(&voter) || self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.skip.entry(voter) {
+                    Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::SkipFallback(_) => {
+                if self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.skip_fallback.entry(voter) {
+                    Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::Finalize(_) => {
+                if self.skip.contains_key(&voter) || self.skip_fallback.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                if let Some(map) = self.notar_fallback.get(&voter) {
+                    if !map.is_empty() {
+                        return Err(AddVoteError::Slash);
+                    }
+                }
+                match self.finalize.entry(voter) {
+                    Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::Genesis(_) => {
+                unimplemented!();
+            }
         }
-        self.votes.push(vote);
-        self.total_stake = self.total_stake.saturating_add(validator_stake);
-        Some(self.total_stake)
-    }
-
-    pub(super) fn votes(&self) -> &[VoteMessage] {
-        &self.votes
     }
 
-    pub(super) fn total_stake(&self) -> Stake {
-        self.total_stake
-    }
-
-    pub(super) fn has_prev_validator_vote(&self, validator_vote_key: &Pubkey) -> bool {
-        self.prev_voted_validators.contains(validator_vote_key)
+    /// Get votes for the corresponding [`VoteType`] and block id.
+    ///
+    /// Depending on the type of of vote, block_id is expected to be Some().
+    fn get_votes(&self, vote_type: VoteType, block_id: Option<&Hash>) -> Vec<VoteMessage> {
```

Comment:
in the end, we seem to only use a reference to the returned `Vec`, so maybe there is a way to do this without cloning, e.g. by returning an iterator here